fs/nfs/nfs4proc.c

```diff
@@ -7771,6 +7778,509 @@ static bool nfs4_xattr_list_nfs4_sacl(struct dentry *dentry)
+ */
+static int nfs4_decode_nfsace4(void **data, size_t *remaining, struct nfs4_ace *ace)
+{
+	u32 *p = *data;
+	u32 length;
+	int error = 0;
+
+	/*
+	 * Per RFC8881 6.2.1 acl the format of nfsace4 is:
+	 * uint32_t type
+	 * uint32_t flag
+	 * uint32_t access_mask
+	 * utf8str_mixed who
+	 *
+	 * where utf8str_mixed is a utf8string.  Per RFC 1832
+	 * this is encoded as a uint32_t length followed by
+	 * string bytes, and then padded by null up to a 4 byte
+	 * boundary.
+	 *
+	 * When it comes to checking remaining, first we will check
+	 * the first four uint32_t (4*4 -> 16 bytes), so that we can
+	 * read the length of the who string.  We do not modify remaining
+	 * unless we are successful in decoding the ACE & advancing thru
+	 * the data buffer.
+	 */
+	#define NFSACE4_HEADER_BYTES (4 * sizeof(u32))
+	if (*remaining < NFSACE4_HEADER_BYTES) {
+		pr_err("NFS: %s buffer not large enough to contain ACE head", __func__);
+		return -EOVERFLOW;
+	}
+	ace->type = ntohl(*(p++));
+	ace->flag = ntohl(*(p++));
+	ace->access_mask = ntohl(*(p++));
+	length = ntohl(*(p++));
+
+	/*
+	 * Now that we have the length, check remaining again.
+	 *
+	 * There could be an argument that we check against:
+	 *     (NFSACE4_HEADER_BYTES + XDR_QUADLEN(length))
+	 * however since we won't actually access any of those trailing
+	 * padding bytes in this call, it doesn't really matter.
+	 */
+	if (*remaining < (NFSACE4_HEADER_BYTES + length)) {
+		pr_err("NFS: %s buffer not large enough to contain ACE who", __func__);
+		return -EOVERFLOW;
+	}
+
+	ace->whotype = nfs4_acl_get_whotype((char *)p, length);
+
+	if (ace->whotype == NFS4_ACL_WHO_NAMED) {
+		/*
+		 * We're just going to support numeric IDs here.
+		 * (it's very expensive to do otherwise)
+		 */
+		unsigned long long id = -1;
+
+		if (numeric_name_to_id((const char *)p, length, &id)) {
+			if (ace->flag & NFS4_ACE_IDENTIFIER_GROUP) {
+				ace->who_gid = make_kgid(&init_user_ns, id);
+				if (!gid_valid(ace->who_gid)) {
+					pr_err("NFS: %s Invalid who_gid: %.*s", __func__,
+					       length, (char *)p);
+					error = -EINVAL;
+				}
+			} else {
+				ace->who_uid = make_kuid(&init_user_ns, id);
+				if (!uid_valid(ace->who_uid)) {
+					pr_err("NFS: %s Invalid who_uid: %.*s", __func__,
+					       length, (char *)p);
+					error = -EINVAL;
+				}
+			}
+		} else {
+			pr_err("NFS: %s Non-numeric who: %.*s", __func__, length, (char *)p);
+			error = -EINVAL;
+		}
+	}
+	/*
+	 * If the whotype was a value other than NFS4_ACL_WHO_NAMED
+	 * (e.g. NFS4_ACL_WHO_OWNER), then no further decoding is
+	 * required (so, no else statement).
+	 */
+
+	if (error >= 0) {
+		/*
+		 * Advance by a multiple of 4 bytes (string + padding).
+		 * Since p is a pointer to u32 we don't need to << 2 the
+		 * output from XDR_QUADLEN.
+		 */
+		p += XDR_QUADLEN(length);
+		*data = p;
+		/* Reduce remaining by initial 4 uint32_t + the string (incl padding) */
+		*remaining -= (NFSACE4_HEADER_BYTES + length);
```

Comment:
"the string (incl padding)" -- where padding came from? Is it part of the length? If not, as it sounds from your comment above, then you must handle it too, or remaining value will be incorrect.  Also size_t is unsigned. Previously you've checked that you have enough space without padding. but here you subtract with padding, that means you may get negative overflow and all the following checks will become useless.

Reply:
Dang.  I wrote one thing in the comment and then did another!  Thanks, good catch.

I've now rectified, but have also made the code more explicit / clearer.  The input buffer that we are processing should contain the string followed by padding (if applicable).

`<length><string of length bytes><padding, if necessary>`